### DISCUSSION TITLE
Form. How to add readonly attribute to text field?

### DISCUSSION THREAD

**Discussion by massimomarazzi** [filamentphp/filament#5463] [Help] [answered]
Hi there, i know that is possible to disable a field with ->disabled().

But disabled fields aren't in POST so it doesn't saved.

I need readonly field because i want to save value, but user cannot edit directly. 
Value of readonly field is a result of calculation  trought other fields.

I also know that i can make textinput disabled() and next add a Hidden field, in this way i have i text field not editabled and i can save field with hidden..... But i think it not elegant solutions and lot of code for nothing... 

Best solution is a add method  ->readonly() like ->disabled()

Or have you any suggestion?

Regards

**Comment by massimomarazzi** [comment, marked as answer]
`->extraInputAttributes(['readonly' => true])`

**Reply by emmanpbarrameda** [reply]
> ->extraInputAttributes(['readonly' => true])

is this possible on 'Select component', sir @massimomarazzi

**Comment by danharrin** [comment]
I'm pretty sure we have `readonly()` already

**Comment by axelitus** [comment]
Don't think so:
![image](https://user-images.githubusercontent.com/732441/229241306-d7e83611-02f0-4103-83c6-eb385951abeb.png)

**Reply by danharrin** [reply]
Try `->readOnly()` instead of `->readonly()`. Or maybe you're on an old Filament version.

**Reply by brentkelly** [reply]
As a FYI to anyone searching, as of the current version (3.2.97) `->readOnly()` is only available on a few form components - e.g. `TextInput` or `Textarea` or any other component that uses `Concerns\CanBeReadOnly`.

It is not available on e.g. `Select` or `Toggle`.

As mentioned below, you can use `->disabled()`, but be aware values will not be dehydrated, possibly resulting in them being cleared in your record on save.

**Comment by axelitus** [comment]
I've check and I'm on Filament v2.17.21. Phpstorm does not detect a readOnly() function for components (in either form):
![image](https://user-images.githubusercontent.com/732441/229291650-6d911550-5e80-48b9-ba38-0c1f881b6c10.png)

Doing a quick search for readonly in filament's vendor directory it shows that there is no function defined:
![image](https://user-images.githubusercontent.com/732441/229291603-003de25f-ba27-4591-80f9-06a11692107c.png)

Maybe it is just implemented in the upcoming v3.x branch?

**Reply by olipiskandar** [reply]
try 
->disabled()

**Reply by njxqlus** [reply]
It's implemented in 3.x Filament version.
```php
 Forms\Components\TextInput::make('slug')
                    ->required()
                    ->maxLength(255)
                    ->readOnly(),
```

**Reply by kombiahrk** [reply]
if you use disabled() values will not pass to form if slug getting values from name field

**Reply by gludovatza** [reply]
@kombiahrk , that's correct. In my example, the creator_id select field is disabled, but I would like to send a default value (user id) to the server. So in the Create[ResourceName].php page I need to add a new function:
```
protected function mutateFormDataBeforeCreate(array $data): array
{
  if( ! isset($data['creator_id']))
    $data['creator_id'] = auth()->user()->id;

  return $data;
}
```
And in the Edit[ResourceName].php page:
```
protected function mutateFormDataBeforeUpdate(array $data): array
{
    if( ! isset($data['creator_id']))
        $data['creator_id'] = auth()->user()->id;

    return $data;
}
```

**Comment by mailtijmen** [comment]
Perhaps helpful to include in the upgrade guides that disabled() no longer passes data and that readonly() should be used.
Yes disabled used to pass data in the previous filament versions. 
I believe I went from 1 to 3 so not sure about version 2
